### DISCUSSION TITLE
PM2 no longer supported, new `gulp dev` command

### DISCUSSION THREAD

**Discussion by tephenavies** [TerriaJS/terriajs#6731] [Announcements]
We've removed pm2 from our dependencies as of terriajs-server@4.0.0 and will no longer ship configuration for running terriajs-server with pm2. We've done this as we no longer use pm2 to run our services in production, instead using kubernetes features to replace much of the functionality that pm2 gave us. Removing pm2 has also removed many transitive dependencies from terriajs-server, and removed the only dependencies that had GPL family licences.

Changes include:

- `npm start` now runs terriajs-server in the foreground
- `npm stop` has been removed
- A new task `gulp dev` has been introduced (in TerriaMap and terriajs) for development. It runs terriajs-server and starts the development watch build. You can pass 1 or more arguments to terriajs-server in one of the following ways, depending on how you run your gulp tasks:
  - `gulp dev --terriajsServerArg port=4000 --terriajsServerArg verbose`
  - `yarn gulp dev --terriajsServerArg port=4000 --terriajsServerArg verbose`
  - `npm run gulp dev -- --terriajsServerArg port=4000 --terriajsServerArg verbose`
- Old aws deploy scripts have been removed

If you wish to continue using pm2 you can install it independently and run e.g.:

```sh
pm2 start ./node_modules/.bin/terriajs-server -i 4 -- --port 4000 --verbose
```

(refer to [pm2 docs](https://pm2.keymetrics.io/docs/usage/quick-start/) for more help).

**Comment by amruthkiran94** [comment]
Hi @steve9164 Thank you for this. I was also wondering if the node, npm and yarn versions described in the documentation still hold for this. Since the documentation (v8) isn't updated to the new commands given above.

**Reply by tephenavies** [reply]
Hi @amruthkiran94. There are no changes to supported versions of node (everything should work with node 14-18), and we still suggest using yarn v1 (specifically yarn v1.18.0 works best when using workspaces to build a custom terriajs with TerriaMap).

I think most of our docs are not affected by this change, as it's still advised to run `yarn gulp && yarn start` when getting started with Terria. This page does need an update: https://docs.terria.io/guide/contributing/development-environment/

**Comment by zoran995** [comment]
Hi @steve9164 I just tried running TerriaJS and TerriaMap on windows machine using proposed approach (`yarn run gulp dev`) but it fails to terriajs-server with 

```
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1124:15)
    at Module._compile (node:internal/modules/cjs/loader:1160:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)
    at Module.load (node:internal/modules/cjs/loader:1074:32)
    at Function.Module._load (node:internal/modules/cjs/loader:909:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:22:47
```

an alternative is to run separately `yarn gulp watch` and `yarn start`, but is there some other option?

**Reply by tephenavies** [reply]
I had intended for `gulp dev` to work on Windows (except for potentially not killing the terriajs-server process in some cases due to differences in signals), but didn't have the time to test it myself.

After looking through a rabbit hole of some other github issues and [this PR](https://github.com/nestjsx/crud/pull/454/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L21-L22) I think Windows doesn't use the `.bin` directory when dependencies are installed. Can you try out https://github.com/TerriaJS/TerriaMap/pull/649 to see if that fixes your problem?

Thanks for letting us know!

**Reply by zoran995** [reply]
It works with those changes, although after playing with it a little bit I will probably stick with starting watch mode and server separately as it's easier to play with sever config when controlled independently. Something that you are probably aware of, but as a side node, if the user starts only a terriajs-server with `yarn gulp terriajs-server` it will just say `Starting 'terriajs-server'...` there is no indication in the console that it is started. Still, maybe that is expected as all logs are directed to the `terriajs-server.log`.
